AP_Motors: Matrix: mixer simplification

IamPete1 opened the pull request:
This is simplification for the main motor mixer, the goal is to make the mixer easier to understand while being mathematically identical. This is more of a what it might look like than a tested suggestion. If this seems like a good idea we can test it is functionally the same using the motors example script. 

- moves the variables declared at the top down to where there used, this means that the handy comment that explains what they do is in the same place the variable is used. This also means a couple can become const.
- ~moves to `if (!motor_enabled[i]) { continue; }` pattern. This means code is not a for and a if loop.~
- removes the unused variable `rp_low` and `rp_high`
- adds a `boost_ratio` helper to simplify the code.
- adds a `const float` intermediate step in the yaw headroom calc
- un-nests two if statements in the `thr_adj` calculation. 
- moves `fabsf` to the denominator in the yaw headroom calc.

lthall commented:
Wow, nice code review and clean up!

I will take some time to check this but thanks for all the effort that went into improving this!!

IamPete1 commented:
Updated the motors test to allow running from the command line in one step and passing  yaw head room and throttle average max values. Then added a shell script that runs the test for a range of those values. This gives 121 files each with a 101871 separate roll/pitch/yaw/throttle input combinations. I have run this on master and again on this re-work for a octo X. There is no change in 4 decimal places of per motor thrust values and no change in the PWM output values.

So I'm fairly happy that there is no change in bahaiour.

However, I have not tested the motor loss case. I'm not sure of the best way to do that, @lthall maybe you have some ideas? We could just set `_thrust_boost` true and let the mixer pick the failed motor, or we could also set `_motor_lost_index` to prescribe the motor that has "failed".

IamPete1 commented:
Rebased and added boost throttle option to the motors example. I have tested octo and quad with the new script, no change in output. It sweeps roll,pitch,yaw and throttle inputs at a range of yaw head room and max average throttle values with and without boost enabled (just under 25 million test points).

lthall commented on libraries/AP_Motors/AP_MotorsMatrix.cpp:
```diff
@@ -200,32 +200,38 @@ uint32_t AP_MotorsMatrix::get_motor_mask()
+    float throttle_avg_max = _throttle_avg_max * compensation_gain;
+
+    // throttle thrust maximum value, 0.0 - 1.0, If thrust boost is active then do not limit maximum thrust
+    const float throttle_thrust_max = _thrust_boost_ratio + (1.0f - _thrust_boost_ratio) * _throttle_thrust_max * compensation_gain;
```
I think we can make this:
const float throttle_thrust_max = boost_ratio(1.0, _throttle_thrust_max * compensation_gain);

IamPete1 replied:
Yep, thanks.

lthall commented on libraries/AP_Motors/AP_MotorsMatrix.cpp:
```diff
@@ -266,30 +273,35 @@ void AP_MotorsMatrix::output_armed_stabilizing()
     // calculate amount of yaw we can fit into the throttle range
     // this is always equal to or less than the requested yaw from the pilot or rate controller
-    float rp_low = 1.0f;    // lowest thrust value
     float rp_high = -1.0f;  // highest thrust value
```
Correct me if I am wrong but I think when I changed the calculation of yaw_alowed on the old line 317 and 319 I removed the need to calculate both rp_low AND rp_high.......
Missed that one didn't I!

IamPete1 replied:
Yep, `rp_high` also removed.

lthall commented on libraries/AP_Motors/AP_MotorsMatrix.cpp:
```diff
@@ -266,30 +273,35 @@ void AP_MotorsMatrix::output_armed_stabilizing()
-            }
+    float yaw_allowed = 1.0f; // amount of yaw we can fit in
+    for (uint8_t i = 0; i < AP_MOTORS_MAX_NUM_MOTORS; i++) {
+        if (!motor_enabled[i]) {
```
What is the advantage of the pattern:
 if (!motor_enabled[i]) {
    continue;
}

over:
 if (motor_enabled[i]) {
     bla bla
}

IamPete1 replied:
It keeps the if small and makes it clear that we don't do anything if not enabled (there is can be nothing after the if and still in the for loop). Also means that the key code is only in a single loop. 

More of a personal preference thing. I have removed. Its not so bad now in any case as we have removed `rp_high` and `rp_low` so the if's are smaller.

lthall replied:
Thanks

lthall commented on libraries/AP_Motors/AP_MotorsMatrix.cpp:
```diff
@@ -266,30 +273,35 @@ void AP_MotorsMatrix::output_armed_stabilizing()
+        // Check the maximum yaw control that can be used on this channel
+        // Exclude any lost motors if thrust boost is enabled
+        if (!is_zero(_yaw_factor[i]) && (!_thrust_boost || i != _motor_lost_index)) {
+            const float roll_pitch_best_throttle = throttle_thrust_best_rpy + _thrust_rpyt_out[i];
```
I think this name is misleading. Maybe thrust_rp_best_throttle or motor_rp_best_throttle.
Given we should be using something like actuator rather than motor and the output may be thrust or torque but not throttle..... I am not going to die on any naming hill.

IamPete1 replied:
Changed to `thrust_rp_best_throttle`

lthall commented on libraries/AP_Motors/AP_MotorsMatrix.cpp:
```diff
@@ -330,30 +346,32 @@ void AP_MotorsMatrix::output_armed_stabilizing()
-        if (motor_enabled[i]) {
-            _thrust_rpyt_out[i] = _thrust_rpyt_out[i] + yaw_thrust * _yaw_factor[i];
+    for (uint8_t i = 0; i < AP_MOTORS_MAX_NUM_MOTORS; i++) {
+        if (!motor_enabled[i]) {
```
I find the other pattern easier to read.....

lthall replied:
less changes too

lthall commented on libraries/AP_Motors/AP_MotorsMulticopter.cpp:
```diff
@@ -857,3 +857,16 @@ bool AP_MotorsMulticopter::arming_checks(size_t buflen, char *buffer) const
     return true;
 }
+
+#if APM_BUILD_TYPE(APM_BUILD_UNKNOWN)
```
Should we have the #if in the .h file too?
I am assuming the answer is no but better to ask a stupid question...

IamPete1 replied:
It would be nice but the current build system doesn't allow it, header files cant "see" the `APM_BUILD_TYPE` define.

lthall reviewed (changes requested):
I really like this!

I don't like the if - continue pattern but I think this is a great improvement.

I suspect we can remove the rp_high now too.

IamPete1 commented on libraries/AP_Motors/AP_MotorsMatrix.cpp:
```diff
@@ -200,32 +200,38 @@ uint32_t AP_MotorsMatrix::get_motor_mask()
+    const float pitch_thrust = (_pitch_in + _pitch_in_ff) * compensation_gain;
+
+    // yaw thrust input value, +/- 1.0
+    float yaw_thrust = (_yaw_in + _yaw_in_ff) * compensation_gain;
```
We set it on 342.

lthall replied:
Sorry, missed that.

IamPete1 commented on libraries/AP_Motors/AP_MotorsMatrix.cpp:
```diff
@@ -200,32 +200,38 @@ uint32_t AP_MotorsMatrix::get_motor_mask()
+    float yaw_thrust = (_yaw_in + _yaw_in_ff) * compensation_gain;
+
+    // throttle thrust input value, 0.0 - 1.0
+    float throttle_thrust = get_throttle() * compensation_gain;
```
Its changed on 238 and 242

lthall replied:
Sorry, missed that.

IamPete1 commented:
Done review fixups, moved away from the `continue` pattern. Just re-running the motor example locally to make sure there is still no change.

lthall reviewed (approved):
Approving assuming the checks show there is no change to the output.

IamPete1 commented:
Local diff just finished, no change in output.
